File: mfiles/draw_snap_pcolor.m
% Draw seismic wavefield snapshot by using pcolor style
% Author:       Yuanhang Huo
% Email:        yhhuo@mail.ustc.edu.cn
% Affiliation:  University of Science and Technology of China
% Date:         2021.05.31

clear all;

% -------------------------- parameters input -------------------------- %
% file and path name
%parfnm='./project/test.json';
%output_dir='./project/output';
%parfnm='/home/zhangw/work/cgfd_cart/13nc/test.json'
%output_dir='/home/zhangw/work/cgfd_cart/13nc/output'
parfnm='/home/zhangw/work/cgfd_ac/00/test.json'
output_dir='/home/zhangw/work/cgfd_ac/00/output'

% which snapshot to plot
id=1;

%-- z slice
subs=[1,1,53];      % start from index '1'
subc=[-1,-1,1];     % '-1' to plot all points in this dimension
subt=[1,1,1];

%-- y slice
%subs=[1,41,1];      % start from index '1'
%subc=[-1,1,-1];     % '-1' to plot all points in this dimension
%subt=[1,1,1];

%-- x slice
%subs=[41,1,1];      % start from index '1'
%subc=[1,-1,-1];     % '-1' to plot all points in this dimension
%subt=[1,1,1];

% variable and time to plot
varnm='Vx';
ns=1;
ne=500;
nt=50;
%ns=2;
%ne=500;
%nt=2;

% figure control parameters
flag_km     = 1;
flag_emlast = 1;
flag_print  = 0;
scl_daspect =[1 1 1];
clrmp       = 'parula';
taut=0.5;
% ---------------------------------------------------------------------- %



% load snapshot data
snapinfo=locate_snap(parfnm,id,'start',subs,'count',subc,'stride',subt,'snapdir',output_dir);
% get coordinate data
[x,y,z]=gather_coord(snapinfo,'coorddir',output_dir);
nx=size(x,1);
ny=size(x,2);
nz=size(x,3);
% coordinate unit
str_unit='m';
if flag_km
   x=x/1e3;
   y=y/1e3;
   z=z/1e3;
   str_unit='km';
end

% figure plot
hid=figure;
set(hid,'BackingStore','on');

% snapshot show
for nlayer=ns:nt:ne
    
    [v,t]=gather_snap(snapinfo,nlayer,varnm,'snapdir',output_dir);
    
    disp([ '  draw ' num2str(nlayer) 'th time step (t=' num2str(t) ')']);
    
    if nx==1
        if flag_emlast
            sid=pcolor((flipud(permute(squeeze(y),[2 1]))), ...
                (flipud(permute(squeeze(z),[2 1]))), ...
                (flipud(permute(squeeze(v),[2 1]))));
        else
            sid=pcolor((permute(squeeze(y),[2 1])), ...
                (permute(squeeze(z),[2 1])), ...
                (permute(squeeze(v),[2 1])));
        end
        xlabel(['Y axis (' str_unit ')']);
        ylabel(['Z axis (' str_unit ')']);
        
    elseif ny==1
        if flag_emlast
            sid=pcolor((flipud(permute(squeeze(x),[2 1]))), ...
                (flipud(permute(squeeze(z),[2 1]))), ...
                (flipud(permute(squeeze(v),[2 1]))));
        else
            sid=pcolor((permute(squeeze(x),[2 1])), ...
                (permute(squeeze(z),[2 1])), ...
                (permute(squeeze(v),[2 1])));
        end
        xlabel(['X axis (' str_unit ')']);
        ylabel(['Z axis (' str_unit ')']);
        
    else
        if flag_emlast
            sid=pcolor((flipud(permute(squeeze(x),[2 1]))), ...
                (flipud(permute(squeeze(y),[2 1]))), ...
                (flipud(permute(squeeze(v),[2 1]))));
        else
            sid=pcolor((permute(squeeze(x),[2 1])), ...
                (permute(squeeze(y),[2 1])), ...
                (permute(squeeze(v),[2 1])));
        end
        xlabel(['X axis (' str_unit ')']);
        ylabel(['Y axis (' str_unit ')']);
    end
    
    set(gca,'layer','top');
    set(gcf,'color','white','renderer','painters');

    % axis image
    % shading
    % shading interp;
    shading flat;
    % colorbar range/scale
    if exist('scl_caxis')
        caxis(scl_caxis);
    end
    % axis daspect
    if exist('scl_daspect')
        daspect(scl_daspect);
    end
    % colormap and colorbar
    if exist('clrmp')
        colormap(clrmp);
    end
    colorbar('vert');
    
    %title
    titlestr=['Snapshot of ' varnm ' at ' ...
              '{\fontsize{12}{\bf ' ...
              num2str((t),'%7.3f') ...
              '}}s'];
    title(titlestr);
    
    drawnow;
    pause(taut);
    
    % save and print figure
    if flag_print==1
        width= 500;
        height=500;
        set(gcf,'paperpositionmode','manual');
        set(gcf,'paperunits','points');
        set(gcf,'papersize',[width,height]);
        set(gcf,'paperposition',[0,0,width,height]);
        fnm_out=[varnm '_ndim_',num2str(nlayer,'%5.5i')];
        print(gcf,[fnm_out '.png'],'-dpng');
    end
    
end

File: pyfiles/v1/draw_snap_pcolor.py
'''
Draw seismic wavefield snapshot by using pcolor style
Author:      Yuanhang Huo
Email:       yhhuo@mail.ustc.edu.cn
Affiliation: University of Science and Technology of China
Date:        2021.06.23
'''

import numpy as np
import matplotlib.pyplot as plt
import imageio
import subprocess
import sys
sys.path.append(".")
from locate_snap import *
from gather_snap import *
from gather_coord import *

# ---------------------------- parameters input --------------------------- #
# file and path name
parfnm= '/home/zhangw/work/cgfd_arc/09/test.json'
snap_dir='/home/zhangw/work/cgfd_arc/09/output'

# which snapshot to plot
id=1

#- z plane
#subs=[1,1,50]       # start from index '1'
#subc=[-1,-1,1]      # '-1' to plot all points in this dimension
#subt=[1,1,1]

#- x plane
subs=[41,1,1]       # start from index '1'
subc=[1,-1,-1]      # '-1' to plot all points in this dimension
subt=[1,1,1]

#- y plane
#subs=[1,41,1]       # start from index '1'
#subc=[-1,1,-1]      # '-1' to plot all points in this dimension
#subt=[1,1,1]

# variable and time to plot
varnm='Vz'
ns=1
ne=500
nt=10

# figure control parameters
# 1
flag_show    = 1
taut         = 0.5
# 2
flag_imgsave = 1
flag_gifsave = 1
figpath      = './fig'
fignm        = 'fd3dsnap.png'
figsize      = [4,4]
figdpi       = 150
# 3
flag_km      = 1
clbtype      = 'seismic'
clbrange     = [None,None]
# ------------------------------------------------------------------------- #



# load snapshot data
snapinfo=locate_snap(parfnm,id,'start',subs,'count',subc,'stride',subt,'snapdir',snap_dir)

# get coordinate data
[x,y,z]=gather_coord(snapinfo,'coorddir',snap_dir)
nx=x.shape[0]
ny=x.shape[1]
nz=x.shape[2]
# coordinate unit
str_unit='m'
if flag_km:
    x=x/1e3
    y=y/1e3
    z=z/1e3
    str_unit='km'

# snapshot show
#plt.figure(dpi=figdpi,figsize=(figsize[0],figsize[1]))
plt.figure()

for nlayer in range(ns,ne+nt,nt):

    v,t=gather_snap(snapinfo,nlayer,varnm,'snapdir',snap_dir)
    print("Drawing " + str(nlayer) + "th time step (t = " + "{:.4f}".format(t) + " s)")

    figtitle="Snapshot of " + varnm + " at " + "{:9.4f}".format(t) + " s"

    if nx == 1:

        #Y=np.squeeze(y).transpose(1,0)
        #Y=np.row_stack((Y,Y[-1,:]))
        #Y=np.column_stack((Y,Y[:,-1]+(Y[0,-1]-Y[0,-2])))
        #Z=np.squeeze(z).transpose(1,0)
        #Z=np.row_stack((Z,Z[-1,:]+(Z[-1,0]-Z[-2,0])))
        #Z=np.column_stack((Z,Z[:,-1]))
        #V=np.squeeze(v).transpose(1,0)
        #V=np.row_stack((V,V[-1,:]))
        #V=np.column_stack((V,V[:,-1]))
        
        Y=np.squeeze(y).transpose(1,0)
        Z=np.squeeze(z).transpose(1,0)
        V=np.squeeze(v).transpose(1,0)

        plt.clf()
        plt.cla()
        plt.pcolor(Y,Z,V,cmap=clbtype,vmin=clbrange[0],vmax=clbrange[1])

        plt.xlabel('Y ' + '(' + str_unit + ')')
        plt.ylabel('Z ' + '(' + str_unit + ')')
        plt.title(figtitle)
        plt.colorbar()
        plt.axis('image')
        if flag_show:
            plt.pause(taut)

    elif ny == 1:
        
        #X=np.squeeze(x).transpose(1,0)
        #X=np.row_stack((X,X[-1,:]))
        #X=np.column_stack((X,X[:,-1]+(X[0,-1]-X[0,-2])))
        #Z=np.squeeze(z).transpose(1,0)
        #Z=np.row_stack((Z,Z[-1,:]+(Z[-1,0]-Z[-2,0])))
        #Z=np.column_stack((Z,Z[:,-1]))
        #V=np.squeeze(v).transpose(1,0)
        #V=np.row_stack((V,V[-1,:]))
        #V=np.column_stack((V,V[:,-1]))

        X=np.squeeze(x).transpose(1,0)
        Z=np.squeeze(z).transpose(1,0)
        V=np.squeeze(v).transpose(1,0)

        plt.clf()
        plt.cla()
        plt.pcolor(X,Z,V,cmap=clbtype,vmin=clbrange[0],vmax=clbrange[1])

        plt.xlabel('X ' + '(' + str_unit + ')')
        plt.ylabel('Z ' + '(' + str_unit + ')')
        plt.title(figtitle)
        plt.colorbar()
        plt.axis('image')
        if flag_show:
            plt.pause(taut)

    else:

        #X=np.squeeze(x).transpose(1,0)
        #X=np.row_stack((X,X[-1,:]))
        #X=np.column_stack((X,X[:,-1]+(X[0,-1]-X[0,-2])))
        #Y=np.squeeze(y).transpose(1,0)
        #Y=np.row_stack((Y,Y[-1,:]+(Y[-1,0]-Y[-2,0])))
        #Y=np.column_stack((Y,Y[:,-1]))
        #V=np.squeeze(v).transpose(1,0)
        #V=np.row_stack((V,V[-1,:]))
        #V=np.column_stack((V,V[:,-1]))

        X=np.squeeze(x).transpose(1,0)
        Y=np.squeeze(y).transpose(1,0)
        V=np.squeeze(v).transpose(1,0)

        plt.clf()
        plt.cla()
        plt.pcolor(X,Y,V,cmap=clbtype,vmin=clbrange[0],vmax=clbrange[1])

        plt.xlabel('X ' + '(' + str_unit + ')')
        plt.ylabel('Y ' + '(' + str_unit + ')')
        plt.title(figtitle)
        plt.colorbar()
        plt.axis('image')
        if flag_show:
            plt.pause(taut)

    if flag_imgsave or flag_gifsave:
        subprocess.call('mkdir -p {}'.format(figpath),shell=True)
        imgnm=fignm[:-(len(fignm.split('.')[-1])+1)]
        imgfmt=fignm.split('.')[-1]
        imgfullnm='{}/{}_timestep_{}.{}'.format(figpath,imgnm,nlayer,imgfmt)
        plt.savefig(imgfullnm)

if flag_gifsave:
    frames=[]
    for nlayer in range(ns,ne+nt,nt):
        imgfullnm='{}/{}_timestep_{}.{}'.format(figpath,imgnm,nlayer,imgfmt)
        frames.append(imageio.imread(imgfullnm))
    imageio.mimsave('{}/{}.gif'.format(figpath,imgnm),frames,'GIF',duration=taut)

if flag_imgsave == 0 and flag_gifsave == 1:
    subprocess.call('rm {}/{}_timestep_*.{}'.format(figpath,imgnm,imgfmt),shell=True)
        

if flag_show:
    plt.show()


